daemon/status.go

```diff
@@ -136,6 +139,109 @@ func (h *getHealthz) Handle(params GetHealthzParams) middleware.Responder {
 	return NewGetHealthzOK().WithPayload(&sr)
 }
 
+type getNodes struct {
+	d *Daemon
+	lock.RWMutex
+	clients map[int64]*clusterNodesClient
+}
+
+func NewGetClusterNodesHandler(d *Daemon) GetClusterNodesHandler {
+	return &getNodes{
+		d:       d,
+		clients: map[int64]*clusterNodesClient{},
+	}
+}
+
+type clusterNodesClient struct {
+	lock.RWMutex
+	*models.ClusterNodeStatus
+}
+
+func (c *clusterNodesClient) NodeAdd(newNode node.Node) error {
+	c.Lock()
+	c.NodesAdded = append(c.NodesAdded, newNode.GetModel())
+	c.Unlock()
+	return nil
+}
+
+func (c *clusterNodesClient) NodeUpdate(oldNode, newNode node.Node) error {
+	c.Lock()
+	c.NodesAdded = append(c.NodesAdded, newNode.GetModel())
+	c.NodesRemoved = append(c.NodesRemoved, oldNode.GetModel())
+	c.Unlock()
+	return nil
+}
+
+func (c *clusterNodesClient) NodeDelete(node node.Node) error {
+	c.Lock()
+	// If the node was added/updated and removed before the clusterNodesClient
+	// was aware of it then we can safely remove it from the list of added
+	// nodes and not set it in the list of removed nodes.
+	found := -1
+	for i, added := range c.NodesAdded {
+		if added.Name == node.Fullname() {
+			found = i
+		}
+	}
+	if found != -1 {
+		c.NodesAdded = append(c.NodesAdded[:found], c.NodesAdded[found+1:]...)
+	} else {
+		c.NodesRemoved = append(c.NodesRemoved, node.GetModel())
+	}
+	c.Unlock()
+	return nil
+}
+
+func (c *clusterNodesClient) NodeValidateImplementation(node node.Node) error {
+	// no-op
+	return nil
+}
+
+func (c *clusterNodesClient) NodeConfigurationChanged(config datapath.LocalNodeConfiguration) error {
+	// no-op
+	return nil
+}
+
+func (h *getNodes) Handle(params GetClusterNodesParams) middleware.Responder {
+	var cns *models.ClusterNodeStatus
+	// If ClientID is not set then we send all nodes and we don't the clusterNodesClient
+	// ID in cache.
+	if params.ClientID == nil {
+		ns := h.d.getNodeStatus()
+		cns = &models.ClusterNodeStatus{
+			Self:       ns.Self,
+			NodesAdded: ns.Nodes,
+		}
+	} else {
+		h.Lock()
+		defer h.Unlock()
+		var clientID int64
+		c, exists := h.clients[*params.ClientID]
+		if exists {
+			clientID = *params.ClientID
+		} else {
+			clientID = rand.Int63()
```

Comment:
I *think* this could return zero, in which case the request you state in the PR could with a very tiny possibility actually fail to allocate a new identity and actually just return a diff from a previously-registered/allocated client '0'?